sycl/doc/extensions/proposed/sycl_ext_oneapi_raw_kernel.asciidoc

```diff
@@ -0,0 +1,513 @@
+namespace syclex = sycl::ext::oneapi::experimental;
+
+struct my_iota;
+static constexpr size_t NUM = 1024;
+
+SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(syclex::range_kernel<1, my_iota>)
+void iota(float start, float *ptr) {
+  // Get the ID of this kernel iteration.
+  size_t id = syclex::this_kernel::get_id();
+
+  ptr[id] = start + static_cast<float>(id);
+}
+
+void main() {
+  sycl::queue q;
+  sycl::context ctxt = q.get_context();
+
+  // Get a kernel bundle that contains the raw kernel "my_iota".
+  auto exe_bndl =
+    sycl::get_kernel_bundle<my_iota, sycl::bundle_state::executable>(ctxt);
+
+  // Get a kernel object for the "my_iota" function from that bundle.
+  sycl::kernel myiota = exe_bndl.get_kernel<my_iota>();
+
+  float *ptr = sycl::malloc_shared<float>(NUM, q);
+  q.submit([&](sycl::handler &cgh) {
+    // Set the values of the kernel arguments.
+    cgh.set_args(3.14f, ptr);
+
+    cgh.parallel_for({NUM}, myiota);
+  }).wait();
+}
+```
+
+
+== Issues
+
+* The front-end team has expressed concern about implementing this syntax
+  because it requires the front-end to recognize the property names
+  `range_kernel`, `nd_range_kernel`, and `single_task_kernel`.
+  This is necessary because the front-end must know that functions decorated
+  with these properties are device code, and it must somehow get the kernel
+  name in order to generate the integration header.
+  Currently, the front-end does not intrinsicly know any of the property names.
+  Rather, the front-end simply passes the properties verbatim into the
+  generated LLVM IR.
+  If we want to avoid teaching the front-end about these property names, we
+  could instead change the syntax for declaring a raw kernel to be like this:
++
+```
+SYCL_EXT_ONEAPI_RANGE_KERNEL(1, my_iota)
+void iota(float start, float *ptr) {
+  // ...
+}
+```
++
+Here, the macro `SYCL_EXT_ONEAPI_RANGE_KERNEL` would expand to two things.
+One part of the expansion would be a C++ attribute that tells the front-end
+that this function is a raw "range" kernel with the given type-name.
+Another part of the expansion would add the same IR attributes as the
+`range_kernel` property (assuming there is even a need to represent this
+information in the IR).
+There are two drawbacks to this approach.
+One is purely aesthetic: we would need to add a new macro to the language
+instead of using our existing property mechanism.
+The other relates to error messages.
+Error messages with macros tend to be worse when the user passes incorrect
+parameters.
++
+This same issue exists also for the extension
+link:https://github.com/intel/llvm/pull/10540[
+sycl_ext_oneapi_virtual_functions] because that extension adds the property
+`indirectly_callable<name>` which must be used to decorate virtual functions
+that can be called from device code.
+Again, the front-end would either need to recognize this property name, or we
+would need to use a new macro like `SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE(name)`
+instead of the property.
```

Comment:
I'm particularly interested in opinions from @intel/dpcpp-specification-reviewers about this first open issue.  How do we feel about the alternate syntax with the macro instead of the property?  I'd like to get this resolved because the decision makes a substantial change in the API.

Reply:
Couldn't we also address this issue by removing the concept of using a type name for the function entirely?

"Named lambdas" have always struck me as a hack anyway.  The use of a type name to describe kernels stems from the fact that kernel function objects are named types, and named lambdas are a way to try and squeeze lambdas into that same box.  If we have an alternative route to naming kernel functions here, I think we should take it.

Reply:
I changed the proposal to use the macro instead of a property in 93504b52ae3e3749a51aade2490236afd4b2be09.  I'm marking this thread as resolved, but feel free to open a new thread if there is more discussion.